Push CBC_BREAKPOINT to the opcode list

polaroi8d opened the pull request:
Add breakpoints:

- active - `(CBC_BREAKPOINT_ENABLED)`  when the additional breakpoint is active, the engine stop running on the breakpoint

- inactive - `(CBC_BREAKPOINT_DISABLED)` and  there is nothing to do when the engine step in the breakpoint

JerryScript-DCO-1.0-Signed-off-by: Levente Orban orbanl@inf.u-szeged.hu

zherczeg commented on jerry-core/parser/js/js-parser.c:
```diff
@@ -1850,6 +1850,10 @@ parser_parse_source (const uint8_t *source_p, /**< valid UTF-8 source code */
 #endif /* !JERRY_NDEBUG */
 
+#ifdef DEBUGGER
+  context.statement_line = 0;
```
When this context.statement_line is updated? I only saw one setting here.

zherczeg commented on jerry-core/parser/js/js-parser-statm.c:
```diff
@@ -1641,6 +1641,20 @@ parser_parse_statements (parser_context_t *context_p) /**< context */
          || context_p->stack_top_uint8 != PARSER_STATEMENT_START)
   {
+#ifdef DEBUGGER
+    if (context_p->line != context_p->statement_line)
```
There is no check whether debugger is enabled during init.

polaroi8d replied:
@zherczeg you think about this one `if (JERRY_CONTEXT (jerry_init_flags) & JERRY_INIT_DEBUGGER)` ?

Commit:
Push CBC_BREAKPOINT to the opcode list

 - Add enabled/disabled breakpoint to the opcode list
 - Add the check whetever the debugger is enable
 - Add 'var' and 'function' statement breakpoints

JerryScript-DCO-1.0-Signed-off-by: Levente Orban orbanl@inf.u-szeged.hu

zherczeg commented on jerry-core/parser/js/js-parser-statm.c:
```diff
@@ -1889,6 +1889,15 @@ parser_parse_statements (parser_context_t *context_p) /**< context */
         break;
       }
     }
+    /* Add breakpoints, after every single line. This breakpoints has an
```
Breakpoints should be added before every line :) When the line starts.

polaroi8d replied:
Okey.

zherczeg commented on jerry-core/parser/js/js-parser-statm.c:
```diff
@@ -1889,6 +1889,15 @@ parser_parse_statements (parser_context_t *context_p) /**< context */
+    *  there is nothing to do when the engine step in the breakpoint.
+    */
+#ifdef JERRY_DEBUGGER
+    parser_emit_cbc (context_p, CBC_BREAKPOINT);
```
First, we need to check whether debugging is enabled, and we should insert NOP-s, which can be turned to debugger at any point by the debugger.

polaroi8d replied:
You think to add a `parser_emit_cbc(context_p, CBC_EXT_NOP);` ?

zherczeg commented on jerry-core/parser/js/byte-code.h:
```diff
@@ -316,8 +316,10 @@
-              VM_OC_BREAKPOINT) \
+  CBC_OPCODE (CBC_BREAKPOINT_ENABLE, CBC_NO_FLAG, 0, \
+              VM_OC_BREAKPOINT_ENABLE) \
+  CBC_OPCODE (CBC_BREAKPOINT_DESABLE, CBC_NO_FLAG, 0, \
```
CBC_BREAKPOINT_ENABLED CBC_BREAKPOINT_DISABLED

zherczeg commented on jerry-core/parser/js/js-parser-statm.c:
```diff
@@ -1656,6 +1656,21 @@ parser_parse_statements (parser_context_t *context_p) /**< context */
     JERRY_ASSERT (context_p->stack_depth == context_p->context_stack_depth);
 #endif /* !JERRY_NDEBUG */
 
+    /* Add breakpoints, before every single line. We have two tpye of breakpoints
```
type

zherczeg commented on jerry-core/parser/js/js-parser-statm.c:
```diff
@@ -1656,6 +1656,21 @@ parser_parse_statements (parser_context_t *context_p) /**< context */
 #endif /* !JERRY_NDEBUG */
 
+    /* Add breakpoints, before every single line. We have two tpye of breakpoints
+    *  one of them the active (enable) others the inactive (desable). In basic
```
enabled, disabled

polaroi8d replied:
Thanks for the review, updated the PR.

zherczeg commented on jerry-core/parser/js/js-parser-statm.c:
```diff
@@ -1656,6 +1656,21 @@ parser_parse_statements (parser_context_t *context_p) /**< context */
+    if (JERRY_CONTEXT (jerry_init_flags) & JERRY_INIT_DEBUGGER)
+    {
+      parser_emit_cbc (context_p, CBC_BREAKPOINT_DISABLED);
+      parser_flush_cbc(context_p);
```
It seems you add a breakpoint before each statement. You should only do it before each line according to the comment above.

zherczeg commented on jerry-core/parser/js/js-parser-internal.h:
```diff
@@ -271,6 +271,10 @@ typedef struct
 #endif /* PARSER_DUMP_BYTE_CODE */
+
+#ifdef JERRY_DEBUGGER
+  parser_line_counter_t last_bp_line;         /**< last line where breakpoint was inserted */
```
last_breakpoint_line

zherczeg commented on jerry-core/parser/js/js-parser-statm.c:
```diff
@@ -1656,6 +1682,29 @@ parser_parse_statements (parser_context_t *context_p) /**< context */
+          && context_p->token.type != LEXER_LEFT_BRACE
+          && context_p->token.type != LEXER_RIGHT_BRACE
+          && context_p->token.type != LEXER_KEYW_VAR
+          && context_p->token.type != LEXER_KEYW_FUNCTION)
```
Add case and default keywords here.

zherczeg commented:
LGTM
